src/Bicep.LangServer/Snippets/SnippetsProvider.cs

```diff
@@ -8,14 +8,20 @@
 using System.Linq;
 using System.Reflection;
 using System.Text;
+using Bicep.Core.Diagnostics;
+using Bicep.Core.Emit;
 using Bicep.Core.Parsing;
+using Bicep.Core.Semantics;
 using Bicep.Core.Syntax;
+using Bicep.Core.TypeSystem.Az;
 using Bicep.LanguageServer.Completions;
 
 namespace Bicep.LanguageServer.Snippets
 {
     public class SnippetsProvider : ISnippetsProvider
     {
+        private Dictionary<string, string> resourceTypeToBodyMap = new Dictionary<string, string>();
```

Comment:
>resourceTypeToBodyMap [](start = 43, length = 21)

Is it possible for the dictionaries to be modified concurrently?

Reply:
And could you add comments to these to explain what they are storing and how that's used?

---
In reply to: [610234912](https://github.com/Azure/bicep/pull/2172#discussion_r610234912) [](ancestors = 610234912)

Reply:
Added comments here:
https://github.com/Azure/bicep/commit/9b3995e29ee139dea8322b474ded7a2b4a34cc94

Reply:
What about the concurrency question? Can multiple threads be modifying the dictionary at the same time?

Reply:
Sorry, I missed this. I don't see a problem with resourceTypeToBodyMap and resourceTypeToDependentsMap as they are modified during Initialize(..). Should happen only once. I am not entirely sure of resourceBodySnippetsCache. I updated to use ConcurrentDictionary to be safe. Let me know if you see any issues or think we need locking mechanism instead. 

https://github.com/Azure/bicep/pull/2172/commits/4fc7f62bf3e8e4bba36946ab621b80b720207ef5

Reply:
Sounds good.